### PR TITLE
zero alloc check: ignore functors and entry functions

### PR DESCRIPTION
**Pull request by gretay-js**
Check of `[@@@zero_alloc all]` annotation already ignores compiler-generated `_entry` functions. It should similarly ignore  compiler-generated functions for functors (see example in the new test). 

There are two changes in this PR:
- In Lambda, adds the attribute `Ignore_assert_all Zero_alloc`  to all functions annotated with attribute `is_a_functor`.
- Similarly, replace the ad-hoc detection of entry functions with `Ignore_assert_all Zero_alloc` at the time the entry function is generated.

### CODE REVIEW

**Comment by lthls** [comment]
You will probably want to also ignore all functions that have the `stub` flag set (though I'm not completely sure for cases like `let f ~x ~y = x + y in let g () = f ~y:0`).

**Comment by lthls** [comment]
> ```ocaml
> let f () =
>   let module M2 = F(M1) in
>   M2.f ()
> ```

I would return `M2.f` directly without applying it, so that we're sure that it's the functor application itself that triggers the check failure instead of the call to `M2.f`.

**Comment by lthls** [comment]
> @lthls @ccasin thank you, I added the test. Regarding stubs, I couldn't construct a test that has an allocation in the stub, is it always just shuffling arguments?

There are multiple kinds of stubs. The functor stubs can allocate directly, but they're already covered by the functor flag. The partial application stubs do not usually allocate themselves, but you can create a problematic example if you use two files:
```ocaml
(* dep.ml *)
let f ~x ~y = (x, y) (* allocates *)

(* main.ml *)
[@@@zero_alloc all]
let g () = Dep.f ~y:0
```
Here after expansions `g` will become:
```ocaml
let g () = (fun[stub] ~x -> Dep.f ~x ~y:0)
```
The stub should be allocated statically, so `g` itself will not allocate, but the stub itself will.

Functions with default parameters also generate stubs:
```ocaml
(* Source code *)
let f x ?(y = (x, x)) () = y
(* After compilation *)
let f_direct x y () = y
let f = fun[stub] x y_opt () ->
  let y = match y_opt with None -> (x, x) | Some y -> y in
  f_direct x y ()
```

Here the stub allocates, and I guess you do want to warn on this.

There are also plans for Flambda2 to generate stubs for function parameter/return unboxing which could allocate (#1248, #1271). So `let[@unbox] f (x : float) = x +. 1.` would generate a raw floating point function that does not allocate, and a stub that allocates for the cases where `f` must be called indirectly. I don't know if you'd expect the `[@@@zero_alloc all]` flag to warn or not in this case.

**Comment by gretay-js** [comment]
> Functions with default parameters also generate stubs:
> let f x ?(y = (x, x)) () = y

In this example and similar ones I've tried, the stub seems to always be inlined according to `-dlambda`, so we get an error which can be turned off with `[@zero_alloc ignore]` as expected.

**Comment by lthls** [comment]
> > Functions with default parameters also generate stubs:
> > let f x ?(y = (x, x)) () = y
> 
> In this example and similar ones I've tried, the stub seems to always be inlined according to `-dlambda`, so we get an error which can be turned off with `[@zero_alloc ignore]` as expected.

The stubs are created in Closure and Flambda, so you wouldn't see them with `-dlambda`.
Although the example I gave you was wrong: because the body is too simple this function isn't split properly.
With the function `let f x ?(y = (x, x)) () = fst y + snd y`, I get the following Cmm output:
```
# With Closure
(function{test.ml:1,6-40} camlTest__f_inner_275
     (x/276: val y/277: val param/278: val)
 (+ (+ (load_mut val y/277) (load_mut val (+a y/277 8))) -1))

(function{test.ml:1,6-40} camlTest__f_267
     (x/269: val *opt*/270: val param/273: val)
 (let
   y/271
     (if (and *opt*/270 1) (alloc{test.ml:1,14-20} 2048 x/269 x/269)
       (load_mut val *opt*/270))
   (app "camlTest__f_inner_275" x/269 y/271 param/273 val)))

# With Flambda
(function{test.ml:1,6-40} camlTest__f_18
     (x/290: val *opt*/289: val param/288: val)
 (let
   y/292
     (if (and *opt*/289 1) (alloc{test.ml:1,14-20} 2048 x/290 x/290)
       (load_mut val *opt*/289))
   (app "camlTest__f_inner_5" x/290 y/292 param/288 val)))

(function{test.ml:1,6-40} camlTest__f_inner_5
     (x/284: val y/283: val param/282: val)
 (let Paddint_arg/285 (load_mut val (+a y/283 8))
   (+ (+ (load_mut val y/283) Paddint_arg/285) -1)))
```
It's not clear from the Cmm output but the non-inner functions are stubs, and they can allocate.
I think that if `f` was annotated with `[@zero_alloc]`, then the check would not trigger despite the fact that calling it as `f 0 ()` would allocate (the stubs have the `Default_check` attribute, regardless of the attribute of the real function).
Although in this case, the fix might be to simply propagate the attribute to the stub: in `ocaml/lambda/simplif.ml`, line 829, `default_stub_attribute` could be replaced with `{ default_stub_attribute with check = attr.check }`.

**Comment by gretay-js** [comment]
> The stubs are created in Closure and Flambda, so you wouldn't see them with -dlambda.

it's a very different kind of stub, and indeed `[@zero_alloc]` check was not sound on it after the previous commits, thanks for the example! I've added it and fixed `Simplif` as you suggested. 

I wonder why other attributes like `[@inline]` are not propagated from `inner` to `stub` in this case as well. It looks like this kind of  "stubs" are always inlined anyway, somehow, even when they are very big. There must be a special case for it in flambda. (Then, if inner has `[@inline]` on it, it will also be inlined and the result would be indistinguishible from the user's point of view, so it's fine not to propagate this annotation and very big stubs are rare I guess.)

**Comment by lthls** [comment]
> I wonder why other attributes like `[@inline]` are not propagated from `inner` to `stub` in this case as well. It looks like this kind of "stubs" are always inlined anyway, somehow, even when they are very big. There must be a special case for it in flambda. (Then, if inner has `[@inline]` on it, it will also be inlined and the result would be indistinguishible from the user's point of view, so it's fine not to propagate this annotation and very big stubs are rare I guess.)

Well, inlining attributes are irrelevant for stubs, as they're supposed to be inlined whenever possible in flambda (we thought that all stubs would be either small or very useful to inline, but it's true that in some cases you could have stubs big enough that they shouldn't be inlined).
I suspect that for other attributes we never really considered whether they should be propagated to stubs or not; for example upstream has the `[@poll]` attribute that can be used to check if a function contains polls, and I think it should be handled as the `check` attribute (i.e. copied into stubs derived from the original function), but apparently nobody ever noticed that it was wrong..

**Comment by gretay-js** [comment]
Unconditional inlining of stubs makes sense, thanks for confirming. 
I'll make a separate PR for `@poll` propagation. 

@lthls can you please approve this PR, if you don't have any other concerns/comments about it?

**Review by lthls — APPROVED** [review]
The changes look fine, although I haven't looked too much at the testsuite changes.
One thing that I noticed is that the printing of the attributes in the `-dlambda` output is a bit hard to parse, but that's not really a concern for this PR.

**Review by xclerc — APPROVED** [review]
